collector/shards.go

```diff
@@ -0,0 +1,177 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	defaultNodeShardLabels = []string{"node"}
+
+	defaultNodeShardLabelValues = func(node string) []string {
+		return []string{
+			node,
+		}
+	}
+)
+
+// Shards information struct
+type Shards struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	nodeShardMetrics  []*nodeShardMetric
+	jsonParseFailures prometheus.Counter
+}
+
+// NodeShard Information per node struct
+type NodeShard struct {
+	node   string
+	shards int64
+}
+
+type nodeShardMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(shards float64) float64
+	Labels func(node string) []string
+}
+
+// NewShards defines Shards Prometheus metrics
+func NewShards(logger log.Logger, client *http.Client, url *url.URL) *Shards {
+	return &Shards{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		nodeShardMetrics: []*nodeShardMetric{
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, "node_shards", "total"),
+					"Total shards per node",
+					defaultNodeShardLabels, nil,
+				),
+				Value: func(shards float64) float64 {
+					return shards
+				},
+				Labels: defaultNodeShardLabelValues,
+			}},
+
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "node_shards", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+	}
+}
+
+// Describe Shards
+func (s *Shards) Describe(ch chan<- *prometheus.Desc) {
+	ch <- s.jsonParseFailures.Desc()
+
+	for _, metric := range s.nodeShardMetrics {
+		ch <- metric.Desc
+	}
+}
+
+func (s *Shards) getAndParseURL(u *url.URL, data interface{}) error {
+	res, err := s.client.Get(u.String())
+	if err != nil {
+		return fmt.Errorf("failed to get from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			_ = level.Warn(s.logger).Log(
+				"msg", "failed to close http.Client",
+				"err", err,
+			)
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	if err := json.NewDecoder(res.Body).Decode(data); err != nil {
+		s.jsonParseFailures.Inc()
+		return err
+	}
+	return nil
+}
+
+func (s *Shards) fetchAndDecodeShards() ([]ShardResponse, error) {
+
+	u := *s.url
+	u.Path = path.Join(u.Path, "/_cat/shards")
+	q := u.Query()
+	q.Set("format", "json")
+	u.RawQuery = q.Encode()
+	u.RawPath = q.Encode()
```

Comment:
I don't think that this belongs here. I'm not sure why the path would come from the query string